sound/soc/sof/virtio-fe.c

```diff
@@ -0,0 +1,734 @@
+			 * Let's update the position now.
+			 */
+			memcpy(&spcm->stream[direction].posn, posn, sizeof(*posn));
+			snd_pcm_period_elapsed(spcm->stream[direction].substream);
+		}
+
+		/* kick back the empty posn buffer immediately */
+		sg_init_one(&sg, posn, sizeof(*posn));
+		virtqueue_add_inbuf(vq, &sg, 1, posn, GFP_KERNEL);
+		virtqueue_kick(vq);
+	}
+}
+
+/*
+ * handle the pos_update, receive the posn and send to up layer, then
+ * resend the buffer to BE
+ */
+static void sof_vfe_psn_handle_rx(struct virtqueue *vq)
+{
+	struct sof_vfe *vfe = vq->vdev->priv;
+
+	schedule_work(&vfe->posn_update_work);
+}
+
+static int sof_vfe_register(struct snd_sof_dev *sdev)
+{
+	return 0;
+}
+
+static int sof_vfe_unregister(struct snd_sof_dev *sdev)
+{
+	return 0;
+}
+
+#define SOF_VFE_FORMATS (SNDRV_PCM_FMTBIT_S16_LE | SNDRV_PCM_FMTBIT_S24_LE | \
+			 SNDRV_PCM_FMTBIT_S32_LE)
+
+struct snd_soc_dai_driver virtio_dai[] = {
+	{
+		.name = "SSP4 Pin",
+		.playback = SOF_DAI_STREAM("ssp4 Tx", 1, 8,
+				SNDRV_PCM_RATE_8000_192000, SOF_VFE_FORMATS),
+		.capture = SOF_DAI_STREAM("ssp4 Rx", 1, 8,
+				SNDRV_PCM_RATE_8000_192000, SOF_VFE_FORMATS),
+	},
+};
+
+static int sof_vfe_run(struct snd_sof_dev *sdev)
+{
+	sdev->boot_complete = true;
+	wake_up(&sdev->boot_wait);
+	return 0;
+}
+
+static void sof_vfe_block_read(struct snd_sof_dev *sdev, u32 bar,
+			       u32 offset, void *dest,
+			       size_t size)
+{
+}
+
+static void sof_vfe_block_write(struct snd_sof_dev *sdev, u32 bar,
+				u32 offset, void *src,
+				size_t size)
+{
+}
+
+static void sof_vfe_ipc_msg_data(struct snd_sof_dev *sdev,
+				 struct snd_pcm_substream *substream,
+				 void *p, size_t sz)
+{
+}
+
+static int sof_vfe_ipc_pcm_params(struct snd_sof_dev *sdev,
+				  struct snd_pcm_substream *substream,
+				  const struct sof_ipc_pcm_params_reply *reply)
+{
+	return 0;
+}
+
+static int sof_vfe_request_topology(struct snd_sof_dev *sdev, const char *name,
+				    struct firmware *fw)
+{
+	struct sof_vfe_ipc_tplg_req rq = {
+		.hdr = {
+			.size = sizeof(rq),
+			.cmd = SOF_IPC_GLB_TPLG_MSG | SOF_IPC_TPLG_VFE_GET,
+		},
+	};
+	struct sof_vfe *vfe = sdev->pdata->vfe;
+	struct sof_vfe_ipc_tplg_resp *partdata = kmalloc(SOF_IPC_MSG_MAX_SIZE,
+							 GFP_KERNEL);
+	size_t part_size = SOF_IPC_MSG_MAX_SIZE - sizeof(partdata->reply),
+		data_size;
+	int ret;
+
+	if (!partdata)
+		return -ENOMEM;
+
+	strncpy(rq.file_name, name, sizeof(rq.file_name));
+
+	mutex_lock(&sdev->ipc->tx_mutex);
+
+	do {
+		size_t to_copy;
+
+		ret = sof_ipc_tx_message_unlocked(sdev->ipc, rq.hdr.cmd,
+						  &rq, sizeof(rq), partdata,
+						  SOF_IPC_MSG_MAX_SIZE);
+		if (ret < 0)
+			goto free;
+
+		data_size = partdata->reply.hdr.size - sizeof(partdata->reply);
+		to_copy = min_t(size_t, data_size, part_size);
+		memcpy(vfe->tplg.data + rq.offset, partdata->data, to_copy);
+		if (!rq.offset)
+			fw->size = data_size;
+		rq.offset += part_size;
+	} while (data_size > part_size);
```

Comment:
Can we ever timeout and abort from here ? we should be able to.

Reply:
@lgirdwood you mean protection against bad data? Ok, will build it in.